### PR TITLE
refactor(params): globalstateless network definition

### PR DESCRIPTION
**Pull request by Wondertan**
Previously we had network being defined as global state in params pkg, this PR changes this so the network is registered on the node itself. This wasn't causing problems for us yet and this is a preemptive measure. Also, the more I think about it(in the context of #552), the more possible issues are jumping into my head and this solution doesn't have those. The main problem was with the way the Swamp environment had to override the network to be Private using a global state. This is now not an issue and each Node keeps track of the network on its own.

### CODE REVIEW

**Review by renaynay — CHANGES_REQUESTED** [review]

File: node/p2p/routing.go [hunk 1]
```diff
@@ -25,10 +25,15 @@ func ContentRouting() routing.ContentRouting {
...
+		if err != nil {
+			return nil, err
+		}
+
 		opts := []dht.Option{
 			dht.Mode(dht.ModeAuto),
-			dht.BootstrapPeers(nparams.BootstrappersInfos()...),
-			dht.ProtocolPrefix(protocol.ID(fmt.Sprintf("/celestia/%s", nparams.GetNetwork()))),
+			dht.BootstrapPeers(bpeers...),
+			dht.ProtocolPrefix(protocol.ID(fmt.Sprintf("/celestia/%s", nparams.DefaultNetwork()))),
```
[the remark this answers is not in the document]
Comment (Wondertan) [reply node/p2p/routing.go:36]: Oh, that's a good catch. Should be a `params.Net` instead, like for the bootstrap peers above.

File: params/default.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+// defaultNetwork defines a default network for the Celestia Node.
+var defaultNetwork = DevNet
+
+// DefaultNetwork returns the network of the current build.
+func DefaultNetwork() Network {
+	return defaultNetwork
+}
+
+// DefaultGenesis reports a hash of a genesis block for the current network.
+func DefaultGenesis() string {
```
[the remark this answers is not in the document]
Comment (Wondertan) [reply params/default.go:14]: Yes, and it doesn't hurt to have it. I believe it will be useful at some point.

File: params/default.go [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+	return genesisList[defaultNetwork] // network is guaranteed to be valid
+}
+
+// DefaultBootstrappersInfos returns address information of bootstrap peers for the node's current network.
+func DefaultBootstrappersInfos() []peer.AddrInfo {
```
[the remark this answers is not in the document]
Comment (Wondertan) [reply params/default.go:19]: Same

File: node/settings.go [hunk 4]
```diff
@@ -14,6 +16,15 @@ import (
 // Option for Node's Config.
 type Option func(*Config, *settings) error
 
+// WithNetwork specifies the Network to which the Node should connect to.
+// WARNING: Use this option with caution and never run the Node with different networks over the same persisted Store.
```
[the remark this answers is not in the document]
Comment (Wondertan) [reply node/settings.go:20]: Too much hassle for the very rare case to happen. Also, this comment is mostly for future devs who will use Node and this option. If they don't read docs, their code will break.

File: node/services/config.go [hunk 5]
```diff
@@ -23,39 +23,44 @@ type Config struct {
...
 		TrustedPeers: make([]string, 0),
 	}
 }
 
-func (cfg *Config) trustedPeers() ([]*peer.AddrInfo, error) {
+func (cfg *Config) trustedPeers(net params.Network) (infos []*peer.AddrInfo, err error) {
 	if len(cfg.TrustedPeers) == 0 {
-		log.Info("No trusted peers given, initializing with default bootstrappers as trusted peers")
-		cfg.TrustedPeers = params.Bootstrappers()
+		log.Info("No trusted peers in config, initializing with default bootstrappers as trusted peers")
```
[the remark this answers is not in the document]
Comment (Wondertan) [reply node/services/config.go:35]: Does not hurt.

File: node/node.go [hunk 6]
```diff
@@ -16,6 +16,8 @@ import (
 	pubsub "github.com/libp2p/go-libp2p-pubsub"
 	"go.uber.org/fx"
 
+	"github.com/celestiaorg/celestia-node/params"
+
```
[the remark this answers is not in the document]
Comment (Wondertan) [reply node/node.go]: Resolved

File: params/default.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (renaynay) [reply params/default.go:19]: https://github.com/celestiaorg/celestia-node/pull/556#discussion_r837651241

File: node/settings.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (renaynay) [reply node/settings.go:20]: ok

File: params/default.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (Wondertan) [reply params/default.go]: That's not that bad. There might be cases where the code is not used within the repo but used outside. Not in this case though, so removing.

File: node/fxutil/options.go [hunk 7]
```diff
@@ -134,7 +134,7 @@ func OverrideSupply(vals ...interface{}) Option {
 				refVal, tp = refVal.Elem(), tp.Elem()
 			}
 
-			if !refVal.IsValid() {
+			if !refVal.IsValid() || refVal.IsZero() {
```
Comment (adlerjohn) [inline node/fxutil/options.go:137]: Why the additional check? Seems orthogonal to the PR.

Comment (Wondertan) [reply node/fxutil/options.go:137]: I extracted it into a separate commit which explains the purpose here.

This makes overriding ignore zero values for built-in non-struct types.

File: node/services/config.go [hunk 8]
```diff
@@ -23,39 +23,44 @@ type Config struct {
...
-// TODO(@Wondertan): We need to hardcode trustedHash hash and one bootstrap peer as trusted.
 func DefaultConfig() Config {
 	return Config{
 		TrustedHash:  "",
 		TrustedPeers: make([]string, 0),
 	}
 }
 
-func (cfg *Config) trustedPeers() ([]*peer.AddrInfo, error) {
+func (cfg *Config) trustedPeers(net params.Network) (infos []*peer.AddrInfo, err error) {
```
Comment (liamsi) [inline node/services/config.go:33]: Not a big fan of named returns btw: https://dave.cheney.net/practical-go/presentations/gophercon-israel.html#_avoid_named_return_values

But no biggie and you can decide to leave this as is or not.

**Review by liamsi — APPROVED** [review]

File: node/services/config.go (hunk 8 above)
Comment (Wondertan) [reply node/services/config.go:33]: @liamsi I am also not a big fan of those. This is the case where it's simple and obvious + avoids an additional and useless `var err error` line by putting it to the return params.
